src/core/pointcloud/expression/qgspointcloudexpressionnode.h

```diff
@@ -0,0 +1,293 @@
+        sipType = sipType_QgsPointCloudExpressionNodeLiteral;
+        break;
+      case QgsPointCloudExpressionNode::ntAttributeRef:
+        sipType = sipType_QgsPointCloudExpressionNodeAttributeRef;
+        break;
+      default:
+        sipType = 0;
+        break;
+    }
+    SIP_END
+#endif
+
+    Q_DECLARE_TR_FUNCTIONS( QgsPointCloudExpressionNode )
+
+  public:
+
+    //! Known node types.
+    enum NodeType
+    {
+      ntUnaryOperator, //!< \see QgsPointCloudExpression::Node::NodeUnaryOperator
+      ntBinaryOperator, //!< \see QgsPointCloudExpression::Node::NodeBinaryOperator
+      ntInOperator, //!< \see QgsExpression::Node::NodeInOperator
+      ntLiteral, //!< \see QgsPointCloudExpression::Node::NodeLiteral
+      ntAttributeRef, //!< \see QgsPointCloudExpression::Node::NodeAttributeRef
+    };
+
+
+    /**
+     * \brief A list of expression nodes.
+     * \ingroup core
+     */
+    class CORE_EXPORT NodeList
+    {
+      public:
+        virtual ~NodeList();
+        //! Takes ownership of the provided node
+        void append( QgsPointCloudExpressionNode *node SIP_TRANSFER ) { mList.append( node ); mNameList.append( QString() ); }
+
+        /**
+         * Returns the number of nodes in the list.
+         */
+        int count() const { return mList.count(); }
+
+        /**
+         * Gets a list of all the nodes.
+         */
+        QList<QgsPointCloudExpressionNode *> list() { return mList; }
+
+        /**
+         * Gets the node at position i in the list.
+         */
+        QgsPointCloudExpressionNode *at( int i ) { return mList.at( i ); }
+
+        /**
+         * Returns a list of names for nodes. Unnamed nodes will be indicated by an empty string in the list.
+         */
+        QStringList names() const { return mNameList; }
+
+        //! Creates a deep copy of this list. Ownership is transferred to the caller
+        QgsPointCloudExpressionNode::NodeList *clone() const SIP_FACTORY;
+
+        /**
+         * Returns a string dump of the expression node.
+         */
+        virtual QString dump() const;
+
+      private:
+        QList<QgsPointCloudExpressionNode *> mList;
+        QStringList mNameList;
+
+
+      public:
+    };
+
+    virtual ~QgsPointCloudExpressionNode() = default;
+
+    /**
+     * Gets the type of this node.
+     *
+     * \returns The type of this node
+     */
+    virtual QgsPointCloudExpressionNode::NodeType nodeType() const = 0;
+
+    /**
+     * Dump this node into a serialized (part) of an expression.
+     * The returned expression does not necessarily literally match
+     * the original expression, it's just guaranteed to behave the same way.
+     */
+    virtual QString dump() const = 0;
+
+    /**
+     * Evaluate this node for the given point and parent.
+     * This will return a cached value if it has been determined to be static
+     * during the prepare() execution.
+     */
+    double eval( QgsPointCloudExpression *parent, int p );
```

Comment:
can we rename `p` to something more meaningful, e.g. pointIndex ?

Reply:
done